crates/lib/src/command/supervisor.rs

```diff
@@ -39,12 +43,75 @@ pub struct Supervisor {
 	ongoing: watch::Receiver<bool>,
 }
 
+#[non_exhaustive]
```

Comment:
One thing I hadn't considered is how to make this constructible outside of this crate. Because of the `#[non_exhaustive]`, it can't be unless a `Default` implementation is provided

that is, this code:

```rust
use watchexec::...::SupervisorBuilder;

SupervisorBuilder {
   // specify all the fields
}
```

won't compile, but this would:

```rust
use watchexec::...::SupervisorBuilder;

SupervisorBuilder {
   // specify all the fields
   ..Default::default()
}
```

However, I'm not sure how to construct a "default" in this case.

Reply:
Also seeing the repetition in this file and that it's not really a builder, perhaps it would be better to call it `SupervisorArgs` and instead of `build()`ing it into a `Supervisor`, have `Supervisor::new(args: SupervisorArgs)`.

Reply:
right...

haha, that is fair, there is a lot of things to think of a good default for in struct! :D

should it be `#[non_exhaustive]` in this case ? it would allow it to be modified later on.. do you think this is likely, or want to keep open that option at least?

we could define a set of args that is definitely required and surely wont change in the future, maybe the candidate being `SupervisorArgs`, and then only have to add defaults for the args that are not included in the `SupervisorArgs` struct?

Reply:
another idea, though maybe quite overkill, could be defining an opaque `Args` type which we can use to get all the required args in `Supervisor::new()`. the other types can then `impl Into<Args>` and this would then only require default values for any fields we add later on!

kind of like:
```rust
pub struct Args {
	errors: Sender<RuntimeError>,
	events: priority::Sender<Event, Priority>,
	commands: Vec<Command>,
	supervisor_id: SupervisorId,
	grouped: bool,
	actioned_events: Arc<[Event]>,
	pre_spawn_handler: HandlerLock<PreSpawn>,
	post_spawn_handler: HandlerLock<PostSpawn>,
}

pub struct RequiredArgs {
	pub errors: Sender<RuntimeError>,
	pub events: priority::Sender<Event, Priority>,
	pub commands: Vec<Command>,
	pub supervisor_id: SupervisorId,
	pub grouped: bool,
	pub actioned_events: Arc<[Event]>,
	pub pre_spawn_handler: HandlerLock<PreSpawn>,
	pub post_spawn_handler: HandlerLock<PostSpawn>,
}

impl Into<Args> for RequiredArgs {
	fn into(self) -> Args {
		let Self {
			errors,
			events,
			commands,
			supervisor_id,
			grouped,
			actioned_events,
			pre_spawn_handler,
			post_spawn_handler,
		} = self;

		Args {
			errors,
			events,
			commands,
			supervisor_id,
			grouped,
			actioned_events,
			pre_spawn_handler,
			post_spawn_handler,
		}
	}
}

impl Supervisor {
	pub fn new(mut args: impl Into<Args>) -> Result<Supervisor, RuntimeError> {
		let args: Args = args.into();

		Supervisor::spawn_with_id(
			args.errors,
			args.events,
			args.commands,
			args.supervisor_id,
			args.grouped,
			args.actioned_events,
			args.pre_spawn_handler,
			args.post_spawn_handler,
		)
	}
}
```
then, later on, if there's enough of a reason we could deprecate `RequiredArgs` for a new args struct that also exposes the new fields ?

does this seem to overkill ?

Reply:
That still makes it breaking to add a required field, unfortunately

Reply:
Alright, keep it as-is, and add a `// FIXME: before release` comment so I remember, and I'll figure it out independently.

Reply:
would it even make it braking if the old `RequiredArgs` was still usable and the `impl Into<Args>` just defaults to some value for that arg, while the new struct, say `RequiredArgsV2` now has that field ? or i may be having a wrong understanding of this too hehe!

Reply:
That would technically work but would also make it quite obnoxious I think.

Reply:
very, very true! do you think this version would be obnoxious already, or it would become more so later one when new versions of `Arg{Version}` are actually added?